kafka/log.go

```diff
@@ -47,33 +48,55 @@ func (h *handle) newLogEvent(cEvent *C.rd_kafka_event_t) LogEvent {
 // Each call to librdkafka times out after timeoutMs. If a call to librdkafka
 // is ongoing when doneChan is closed, the function will wait until the call
 // returns or times out, whatever happens first.
-func (h *handle) pollLogEvents(toChannel chan LogEvent, timeoutMs int, doneChan chan bool) {
+func (h *handle) pollLogEvents(toChannel chan LogEvent, termChan chan bool) {
 	for {
-		select {
-		case <-doneChan:
-			return
+		// See if there's anything ready-to-go without polling
+		cEvent := C.rd_kafka_queue_poll(h.logq, C.int(0))
 
-		default:
-			cEvent := C.rd_kafka_queue_poll(h.logq, C.int(timeoutMs))
-			if cEvent == nil {
-				continue
-			}
+		if cEvent == nil {
```

Comment:
Since this logic is occurring in a couple of places, maybe move it to its own function, eventPollSingle() or something, that performs the immediate and then blocking polls.

Reply:
Yeah I'll factor something out here.

Reply:
Looking at this more carefully - doing this will have a performance impact. The log poller just does a straight-up `rd_kafka_queue_poll`, but the event poller does a call to `_rk_queue_poll` instead which makes some librdkafka calls to fill out message structure info if the returned event is a `RD_KAFKA_EVENT_FETCH`.

Options here would be:
 
1. Leave it how it is, with the duplicated code
2. Change the log poller to also call `_rk_queue_poll`
3. Change the event poller to call `rd_kafka_queue_poll` directly (and take the performance hit of then needing to subsequently make another cgo call in the message case).
4. Make your proposed `eventPollsingle()` generic by allowing one to pass in a callback function.

I think i'm going to go with option 2 here, but let me know if you feel differently.